tests/integration/extension/src/background.js

```diff
@@ -25,4 +28,15 @@ async function pageDataListener(pageData) {
 webScience.pageManager.onPageVisitStart.addListener(pageVisitStartListener);
 webScience.pageManager.onPageVisitStop.addListener(pageVisitStopListener);
 
-webScience.pageNavigation.onPageData.addListener(pageDataListener, { matchPatterns: ["<all_urls>"] });
\ No newline at end of file
+webScience.pageNavigation.onPageData.addListener(pageDataListener, { matchPatterns: ["<all_urls>"] });
+
```

Comment:
side note unrelated to this PR: do we need to worry about these `addListener` and `registerContentScripts` functions being called multiple times due to restarts of the service worker in mv3?

i.e. with mv3 we need to start ensuring our background scripts are "idempotent", right?

Reply:
That's a good question, I'll look into this. It's also an issue with updates, but service worker restarts are much more frequent obviously :)

Reply:
So `registerContentScripts` will fail if we attempt to re-use an ID, so I think those are fine.

I think in the extension update case I brought up the browser will tear down existing listeners, so that's probably not really an issue. They are infrequent enough (and if it is a problem, it's a pre-existing problem) so I'm going to set that aside but it's worth looking into.

I don't think the background script is re-initialized every time it's woken up, it works more like this:

1. browser evaluates and runs the service worker on extension startup
2. service worker adds a `stateChangeCallback` listener in rally SDK
3. service worker initializes Glean, and the script finishes

At some point, the `stateChangeCallback` is called which puts the extension into the `run` state, and it registers some additional listeners. This is where the `WebScience` listeners and any others are added.

The "waking up" that the service worker does is to do things like run callbacks from these already-registered listeners; I don't believe that it does steps 1-3 again until the extension is restarted. Generally an extension restart happens on browser restart or on extension update (which, unless we explicitly delay it with [runtime.onUpdateAvailable](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onUpdateAvailable), will happen immediately and restartlessly). `chrome://extension` developer mode and `about:debugging` in Firefox can force a restart too.

Reply:
ok yes this is making some sense to me. Although for now I'm still trying to wrap my head around the non-Rally case, since we're looking at the WebScience test extension.

I'm realizing I maybe have a fundamental misunderstanding about how service workers work. The top-level code, that's only run once on extension start up? And then subsequent "wakes" only happen if the top-level code registered listeners, and the listener gets triggered? In which case, I think my concern is moot because the top-level code is not run again when the service worker awakes.

BUT what confused me is language in, for example, [this docs section](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/#event_listeners) for Chrome, which says "When a service worker is terminated, so are the event listeners associated with it." Which conflicts with my understanding above.

I've approved the PR, but I do still want to understand this better for future.

Reply:
OK I tested it and I think you're right @aaga - it does look like the service worker gets re-evaluated. As long as the listeners are torn down first I don't *think* it should matter (and content scripts shouldn't get re-registered due to the ID). Nice find, thanks!